openedx/core/djangoapps/discussions/tasks.py

```diff
@@ -51,37 +51,48 @@ def update_discussions_settings_from_course(course_key: CourseKey) -> CourseDisc
     provider_type = discussions_config.provider_type
 
     def iter_discussable_units():
-        subsections = store.get_items(course_key, qualifiers={"category": "sequential"})
         # Start at 99 so that the initial increment starts it at 100.
         # This leaves the first 100 slots for the course wide topics, which is only a concern if there are more
         # than that many.
         idx = 99
-        for subsection in subsections:
-            section = store.get_item(subsection.parent)
-            for unit in subsection.get_children():
-                # Increment index even for skipped units so that the index is more stable and won't change
-                # if settings change, only if a unit is added or removed.
-                idx += 1
-                # If unit-level visibility is enabled and the unit doesn't have discussion enabled, skip it.
-                if unit_level_visibility and not getattr(unit, "discussion_enabled", False):
+        log.info(f"[DEBUG INF-291] Unit-level visibility enabled: {unit_level_visibility}")
+        for section in course.get_children():
+            if section.location.block_type != "chapter":
+                continue
+            for subsection in section.get_children():
+                if subsection.location.block_type != "sequential":
                     continue
-                # If the unit is in a graded section and graded sections aren't enabled skip it.
-                if subsection.graded and not enable_graded_units:
-                    continue
-                # If the unit is an exam, skip it.
-                if subsection.is_practice_exam or subsection.is_proctored_enabled or subsection.is_time_limited:
-                    continue
-                yield DiscussionTopicContext(
-                    usage_key=unit.location,
-                    title=unit.display_name,
-                    group_id=None,
-                    ordering=idx,
-                    context={
-                        "section": section.display_name,
-                        "subsection": subsection.display_name,
-                        "unit": unit.display_name,
-                    },
-                )
+                for unit in subsection.get_children():
+                    if unit.location.block_type != 'vertical':
+                        continue
+                    log.info(f"[DEBUG INF-291] Processing unit: {unit.location}")
+                    # Increment index even for skipped units so that the index is more stable and won't change
+                    # if settings change, only if a unit is added or removed.
+                    idx += 1
```

Comment:
@xitij2000 Shouldn't this happen before the `vertical` check?

```py
idx += 1
if unit.location.block_type != 'vertical':
    continue
```
Based on the subsequent `if` blocks and the `continue` statements, I can see that "skipped units" has a discussion specific meaning. So maybe the code as it is correct in the sense that - "vertical" blocks that are skipped are also counted. If that's the case, can we have the comment explicitly mention this to avoid confusion?

Reply:
I'm in two minds about this. If the block type isn't vertical then that means the course is using a custom hierarchy or some other block that's unsupported. I think in that case, it's fair to not count it since it can never really be counted.

Reply:
I see you point and agree as well.